src/dparse/strings.d

```diff
@@ -0,0 +1,694 @@
+
+		case 'x':
+			if (!requireMinLength(3))
+				break Loop;
+			char a = input[escape + 2];
+			char b = input[escape + 3];
+			if (!a.isHexDigit || !b.isHexDigit)
+			{
+				errorInvalidCharacter(escape + 4);
+				break;
+			}
+			ret ~= cast(char)(a.parseHexChar << 4 | b.parseHexChar);
+			start = escape + 4;
+			break;
+		case 'u':
+			if (!requireMinLength(1 + 4))
+				break Loop;
+			parseUnicode(4);
+			break;
+		case 'U':
+			if (!requireMinLength(1 + 8))
+				break Loop;
+			parseUnicode(8);
+			break;
+		case '0': .. case '7':
+			int length = 1;
+			foreach (n; 2 .. 4)
+			{
+				if (escape + 1 + n > input.length)
+					break;
+				char c = input[escape + n];
+				if (c >= '0' && c <= '7')
+					length = n;
+				else
+					break;
+			}
+			int c = input[escape + 1 .. escape + 1 + length].to!int(8);
+			ret ~= cast(char) c;
+			start = escape + 1 + length;
+			break;
+		case '&':
+			auto end = input.indexOf(';', escape + 2);
+			if (end == -1)
+				errorInvalidCharacter(input.length);
+			else
+				ret ~= input[escape .. start = end + 1];
+			break;
+		default:
+			errorInvalidCharacter(escape + 1);
+			break;
+		}
+
+		escape = input.indexOf('\\', start);
+	}
+	ret ~= input[start .. $];
+	return ret.data;
+}
+
+unittest
+{
+	assert(unescapeDoubleQuotedContent(`hello world`) == "hello world");
+	assert(unescapeDoubleQuotedContent(`hello\nworld`) == "hello\nworld");
+	assert(unescapeDoubleQuotedContent(`hello\tworld`) == "hello\tworld");
+	assert(unescapeDoubleQuotedContent(`hello\u200bworld`) == "hello\u200bworld");
+	assert(unescapeDoubleQuotedContent(`hello \"\\ok`) == "hello \"\\ok");
+}
+
+private string parseHexStringContent(
+	InvalidEscapeAction invalidEscapeAction = InvalidEscapeAction.error
+)(
+	string input
+)
+{
+	if (!input.length)
+		return input;
+
+	auto ret = appender!string;
+	ret.reserve(input.length / 3);
+	char buf;
+	foreach (i, char c; input)
+	{
+		if (c.isWhite)
+			continue;
+
+		if (!c.isHexDigit)
+		{
+			final switch (invalidEscapeAction)
+			{
+			case InvalidEscapeAction.keep:
+				if (buf != char.init)
+				{
+					ret ~= buf;
+					buf = char.init;
+				}
+				ret ~= c;
+				break;
+			case InvalidEscapeAction.skip:
+				break;
+			case InvalidEscapeAction.error:
+				throw new ConvException("Invalid hex character at index "
+						~ i.to!string);
+			}
+		}
+		else
+		{
+			if (buf == char.init)
+			{
+				buf = c;
+			}
+			else
+			{
+				ret ~= cast(char)(buf.parseHexChar << 4 | c.parseHexChar);
+				buf = char.init;
+			}
+		}
+	}
+
+	if (buf != char.init)
+	{
+		final switch (invalidEscapeAction)
+		{
+		case InvalidEscapeAction.keep:
+			ret ~= buf;
+			break;
+		case InvalidEscapeAction.skip:
+			break;
+		case InvalidEscapeAction.error:
+			throw new ConvException("Unterminated hex character at end of string");
+		}
+	}
+
+	return ret.data;
+}
+
+private int parseHexChar(char c)
+in
+{
+	assert(c.isHexDigit);
+	assert('a' > 'A' && 'A' > '0'); // just checking that ASCII doesn't suddenly change
```

Comment:
That's very unlikely.

Reply:
well it does make the intent in the function body clearer (not having maximum checks for the character values) and shows my assumption about the characters, which maybe isn't clear to someone or someone doesn't know the ASCII ordering out the top of their head.

Both because this is an assert (in an `in` block too) and because this is evaluatable at compile time, this should be removed (at least in release mode), so there is no overhead to this.